src/lib/bootstrap.ts

```diff
@@ -119,7 +118,9 @@ function handleHTTPS(app, configPath, config) {
       httpsOptions = assign(httpsOptions, {
         key: fs.readFileSync(config.https.key),
         cert: fs.readFileSync(config.https.cert),
-        ca: fs.readFileSync(config.https.ca),
+        ...(config.https.ca && {
```

Comment:
we do support https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-7.html#optional-chaining optional chaining. Your code will look way safer if you have something like `config?.https?.ca` 👍

Reply:
Thanks for your feedback! I don't mind adding optional chaining, although I have few questions I'd like to discuss:
- if we add optional chaining on this line, should we also add to every other in this function?
- private `handleHTTPS` would only execute when `config.https` [is truthy](https://github.com/verdaccio/verdaccio/pull/1853/files/9e432714d0829f617af0593fee9a855ff6f53b35#diff-aff38a992121eaf783dc7a02739c4125L51)

My personal opinion for the proper way to make this function totally "typescripted" would be:
1. define types in monorepo([PR](https://github.com/verdaccio/monorepo/pull/368))
2. make this function like:
```typescript
function handleHTTPS(app: express.Application, configPath: string, config: ConfigWithHttps): https.Server {
  try {
    let httpsOptions = {
      secureOptions: constants.SSL_OP_NO_SSLv2 | constants.SSL_OP_NO_SSLv3, // disable insecure SSLv2 and SSLv3
    };

    const keyCertConfig = config.https as HttpsConfKeyCert;
    const pfxConfig = config.https as HttpsConfPfx;

    // https must either have key and cert or a pfx and (optionally) a passphrase
    if (!(pfxConfig || keyCertConfig)) {
      logHTTPSWarning(configPath);
    }

    if (pfxConfig) {
      const { pfx, passphrase } = pfxConfig;
      httpsOptions = assign(httpsOptions, {
        pfx: fs.readFileSync(pfx),
        passphrase: passphrase || '',
      });
    } else if (keyCertConfig) {
      const { key, cert, ca } = keyCertConfig;
      httpsOptions = assign(httpsOptions, {
        key: fs.readFileSync(key),
        cert: fs.readFileSync(cert),
        ...(ca && {
          ca: fs.readFileSync(ca),
        }),
      });
    }
    return https.createServer(httpsOptions, app);
  } catch (err) {
    // catch errors related to certificate loading
    logger.logger.fatal({ err: err }, 'cannot create server: @{err.message}');
    process.exit(2);
  }
}
```

Reply:
> Thanks for your feedback! I don't mind adding optional chaining, although I have few questions I'd like to discuss:
> 
> * if we add optional chaining on this line, should we also add to every other in this function?
> * private `handleHTTPS` would only execute when `config.https` [is truthy](https://github.com/verdaccio/verdaccio/pull/1853/files/9e432714d0829f617af0593fee9a855ff6f53b35#diff-aff38a992121eaf783dc7a02739c4125L51)
> 
> My personal opinion for the proper way to make this function totally "typescripted" would be:
> 
> 1. define types in monorepo([PR](https://github.com/verdaccio/monorepo/pull/368))
> 2. make this function like:
> 
> ```ts
> function handleHTTPS(app: express.Application, configPath: string, config: ConfigWithHttps): https.Server {
>   try {
>     let httpsOptions = {
>       secureOptions: constants.SSL_OP_NO_SSLv2 | constants.SSL_OP_NO_SSLv3, // disable insecure SSLv2 and SSLv3
>     };
> 
>     const keyCertConfig = config.https as HttpsConfKeyCert;
>     const pfxConfig = config.https as HttpsConfPfx;
> 
>     // https must either have key and cert or a pfx and (optionally) a passphrase
>     if (!(pfxConfig || keyCertConfig)) {
>       logHTTPSWarning(configPath);
>     }
> 
>     if (pfxConfig) {
>       const { pfx, passphrase } = pfxConfig;
>       httpsOptions = assign(httpsOptions, {
>         pfx: fs.readFileSync(pfx),
>         passphrase: passphrase || '',
>       });
>     } else if (keyCertConfig) {
>       const { key, cert, ca } = keyCertConfig;
>       httpsOptions = assign(httpsOptions, {
>         key: fs.readFileSync(key),
>         cert: fs.readFileSync(cert),
>         ...(ca && {
>           ca: fs.readFileSync(ca),
>         }),
>       });
>     }
>     return https.createServer(httpsOptions, app);
>   } catch (err) {
>     // catch errors related to certificate loading
>     logger.logger.fatal({ err: err }, 'cannot create server: @{err.message}');
>     process.exit(2);
>   }
> }
> ```

> if we add optional chaining on this line, should we also add to every other in this function?

Up to you, feel free to moderniz the code using the latest features, the idea is you can use the latest available stuff :) 

> My personal opinion for the proper way to make this function totally "typescripted" 

Agreed, I merged you PR for types.